### ISSUE TITLE
Never show full desktop entry names on Desktop

### ISSUE THREAD

**Issue by tsujan** [lxqt/pcmanfm-qt#996]
Even when the corresponding option is checked in pcmanfm-qt's Preferences.

I don't know if this is a good change; opened this issue for a discussion.

**Comment by agaida** [comment]
ever show the filenames except of desktop and Applications

**Comment by tsujan** [comment]
> ever show the filenames except of desktop and Applications

I think `Applications` won't be touched automatically.

So, by Desktop we mean the real Desktop and not just the folder desktop that may be opened in a window. Right?

**Comment by agaida** [comment]
![](https://i.imgur.com/NR54VeW.png)

**Comment by agaida** [comment]
![](https://i.imgur.com/Ggg5X3h.png)
vs
![](https://i.imgur.com/01Pp1fq.png)
in a filesystem i consider the second one as pure horror. On a desktop it might be ok.

**Comment by tsujan** [comment]
> A filemanager displays files per default.

The only thing I have problem with is "per default". I like the current default ;) It's like saying that a file manager shouldn't show image/video thumbnails because, in that way, it wouldn't show their icons as files. Or more: pcmanfm-qt should only show an ugly desktop-file icon for all desktop files, like Nautilus does. I can accept neither.

**Comment by agaida** [comment]
![](https://i.imgur.com/L6Iu1vn.png)
i don't like it in a file manager
![](https://i.imgur.com/1GgyA7I.png)
i don't like it too
![](https://i.imgur.com/Tp6ukt5.png)
i like that one for a file manager

**Comment by agaida** [comment]
And maybe: Take the option away from the pcmanfm-qt settings and place it in desktop settings - would make sense

**Comment by tsujan** [comment]
This is what I have no problem with:

If the user checks `Preferences → Display → Always show full names` and restarts pcmanfm-qt, then all desktop entry names will be their file names (vs. display names), except for desktop entries on Desktop (not the folder). The icons should also remain as they are.

It's a matter of taste. Different Desktop entries with the same string after `Name=` aren't usual (and that's a bad practice). The user has an option here. Taking away that option isn't good to me.

**Comment by agaida** [comment]
so let the scope of this option only be the Desktop - done

**Comment by agaida** [comment]
and that would solve another problem - if the user open the "real" Desktop in the filemanager he would see the filenames - without switching and restarting the Desktop - this would be my favorite outcome

**Comment by tsujan** [comment]
Agreed.

BTW, image many users that will curse us if we make a backward-incompatible change by removing an option and changing a visible thing ;) We had that problem with "trust emblem" but the number of people who wanted it (implicitly) was considerable -- or, at least, it seemed to me so.

**Comment by agaida** [comment]
i guess the most users will not notice it - as it will not change the default behaviour

**Comment by tsujan** [comment]
That's tight about our final agreement :)

**Comment by agaida** [comment]
Good night. And i found another intresting thing in Nautilus that i don't expected there - but will leave that for tomorrow.

**Comment by tsujan** [comment]
Good morning!

@agaida, I investigated it. Please read this long explanation; it's somehow important:

*Desktop as Desktop can't be distinguished from Desktop as folder, as far as item texts are concerned.* The reason is simple: we use a `CachedFolderModel`, so that the model that's used inside pcmanfm-qt's window is the same as the one that's used on Desktop — and item texts are determined by that model. That's nice because it minimizes resource usage, among other things.

Therefore, if we enforce display names *only* on Desktop, once the user opens it as a folder inside a pcmanfm-qt window, the desktop entries on Desktop will change their names to their real file names and remain so, because the corresponding setting will be toggled by the pcmanfm-qt window (in `TabPage::chdir`).

A separate folder model for Desktop solves this problem (I tested it) but

 1. Theoretically, it can use a little more resources, although the difference is quite negligible in practice.
 2. It interferes with graying-out of cut files: if a file is cut inside a window, it won't be grayed out on Desktop, and conversely. But that may show that our method of dealing with cut files isn't efficient (I think it isn't and should be changed at some point).

In short:

If a separate folder model isn't created for Desktop (= current code), then *both* Desktop as Desktop *and* Desktop as a folder should show display names, which isn't good. That doesn't happen with a separate folder model for Desktop but then, the visible state of cust files on Desktop won't be synchronized with its folder inside a window and we might have a very negligible increase in resource usage (only when desktop is opened in a window and only theoretically).

Also note that this isn't related to our previous disagreement, which was just about options and defaults.

Now, the difficult question: What do you think?

**Comment by agaida** [comment]
All for a separate folder model for Desktop:  
re 1: doesn't really matter, one can't have and omelette without breaking eggs
re 2: we can live with - it's a corner case, as long we can make sure that both views show the same content after the action is finished.

You see, i changed my mind a bit in our discussion, basically your point about file managers and the to be expected behaviour of a file manager. And we had the discussion before with far more participants: Is it really clever to handle Desktop and file management in the same process - and this will come up again and again and bring me back to the very start of LXQt.

In former days the Desktop painting capabilities of pcmanfm-qt was so poor and buggy that i decided to draw the Desktop with pcmanfm and use pcmanfm-qt only as file manager - it was for our first release, the live iso should be still available :smile:

Ok, we would use a bit more ressources, but both things - different model or two processes are viable options.

**Comment by agaida** [comment]
before i forget - target is not the upcoming release, but +1 - so we should take our time to test and discuss

**Comment by tsujan** [comment]
> target is not the upcoming release

Definitely.

> All for a separate folder model for Desktop

OK. I'll do more experiments and tests. I should see, for example, how much CPU time our thumbnailer takes when a video file is being downloaded to Desktop, while the Desktop folder is also open inside a tab. Our code for cut files isn't efficient either; etc. It'll take some time... But the idea is correct: Desktop should always show the display names of desktop entries --  like menus do.

**Comment by palinek** [comment]
> Desktop as Desktop can't be distinguished from Desktop as folder, as far as item texts are concerned. The reason is simple: we use a CachedFolderModel, so that the model that's used inside pcmanfm-qt's window is the same as the one that's used on Desktop — and item texts are determined by that model. 

Just a humble note: This sounds plain wrong. The model should definitely not enforce any visual presentation. Model should hold the data and visual presentation should be based on view, which renders particular model. If this is not true here, it is architecturally wrong and should be fixed.

**Comment by tsujan** [comment]
A model enforces its data, a part of which is `Qt::DisplayRole`, which is item text in this case.

Nothing is wrong here.

**Comment by palinek** [comment]
But there is no need for another model with the same data. How about changing the presenatation with just a simple delegate which just overrides the `initStyleOption()` ...  https://doc.qt.io/qt-5/qstyleditemdelegate.html#initStyleOption

Just overriding the `option->text`

**Comment by tsujan** [comment]
That can't be done consistently. Everything is based on a model that provides data and a view/delegate that shows them in a certain way (icon-above-text, text-beside-icon, grayed out icon, etc.).

**Comment by palinek** [comment]
If you/we can set e.g. specific margins for `FolderItemDelegate` in `DesktopWindow::updateFromSettings()` why can't we set something like *setDisplayRole* and then use this value in `initStyleOption()` (which just calls parents `initStyleOption()`  and then overrides `option->text`. The `FolderModel::data()` can return different (or the same) value for (new) **DesktopDisplayRole**

**Comment by tsujan** [comment]
Of course, we can change margins and several other things. They're about *how* data are shown, not about which data are shown.

We could override `option->text` too but, this time, we would contradict the model-delegate-view architecture, which would result in various bugs.

**Comment by palinek** [comment]
> They're about how data are shown, not about which data are shown.

Exactly. And the delegate will be presenting the same data item, just some other piece of it, that it needs.

**Comment by tsujan** [comment]
All pieces of data should be already in `option`; adding anything else would result in bugs. Visual parameters aren't "data" in this sense. See https://doc.qt.io/qt-5/model-view-programming.html

**Comment by tsujan** [comment]
Anyway, I *think* that I've fixed the cut problem mentioned above (and more), so that a separate model for desktop should work fine.

**Comment by palinek** [comment]
> All pieces of data should be already in option; 

All pieces of data are initially fed in `QStyledItemDelegate::initStyleOption()` into `option`. Overriding it in (new) `DestkopItemDelegate::initStyleOption()` is the right place IMO.

**Comment by tsujan** [comment]
I *can* use a complex system in a way it isn't intended but I cannot predict what will happen — actually I can predict because I've seen and fixed so many examples of such bugs in LXQt, including pcmanfm-qt and libfm-qt: the least you could expect is styling problems here and there. `initStyleOption()` is just the tip of the iceberg.

I really prefer not to use libfm-qt as a testing ground.

**Comment by tsujan** [comment]
@agaida 
After fixing cut files (made a [PR](https://github.com/lxqt/libfm-qt/pull/460)) and using a separate desktop folder model (no PR yet), I saw that changing the default would cause no problem anymore; nothing would change for current users.

I haven't encountered any problem so far; thumbnail creation is OK and there's no extra CPU usage.

![screenshot](https://user-images.githubusercontent.com/981076/63455813-bf7d2680-c462-11e9-8f1a-a3c6a5a5756e.png)

**Comment by agaida** [comment]
Sounds good
